src/app/icd/ICDManager.cpp

```diff
@@ -228,5 +254,74 @@ void ICDManager::OnTransitionToIdle(System::Layer * aLayer, void * appState)
     pIcdManager->mStateObserver->OnTransitionToIdle();
 }
 
+bool ICDManager::EnsureChipStackLock()
+{
+    // TODO implement platform agnostic lock creation/mechanism that can be used for ICD without locking the chip stack
+    bool requireLock = false;
+
+#if CHIP_STACK_LOCK_TRACKING_ENABLED
+    requireLock = !DeviceLayer::PlatformMgr().IsChipStackLockedByCurrentThread();
+#endif
+
+    if (requireLock)
+    {
+        DeviceLayer::PlatformMgr().LockChipStack();
```

Comment:
Yeah, this is a really weird pattern.  We're basically re-implementing a re-entrant lock here, which is usually hard to reason about.

What should be going on instead is that we should require the lock be held for all our public APIs, and leave synchronization to the consumer.  Who should know which thread context they are running on and hence how to call the API.

Reply:
Left the responsibility to the icdNotifier user to ensure that its calls are done from the chip task context or with the lock held